### PR TITLE
NATS Source, JetStream and Pub/Sub iterators

### PR DESCRIPTION
**Pull request by alyakimenko**
### Description

Implement NATS Source with JetStream and Pub/Sub iterators.

### Quick checks:

- [x] I have followed the [Code Guidelines](https://github.com/ConduitIO/conduit/blob/main/docs/code_guidelines.md).
- [x] There is no other [pull request](https://github.com/ConduitIO/conduit-connector-s3/pulls) for the same update/change.
- [x] I have written unit tests.
- [x] I have made sure that the PR is of reasonable size and can be easily reviewed.

### CODE REVIEW

File: source/jetstream/iterator.go [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+		return nil, fmt.Errorf("get jetstream context: %w", err)
+	}
+
+	consumerOpts, err := getConsumerOptions(params)
+	if err != nil {
+		return nil, fmt.Errorf("get consumer options: %w", err)
+	}
+
+	messages := make(chan *nats.Msg, params.BufferSize)
+	subscription, err := jetstream.ChanSubscribe(params.Subject, messages, consumerOpts...)
```
Comment (hariso) [inline source/jetstream/iterator.go:69]: Similar question to the one we had in NATS Pubsub: can we handle async errors?

File: source/jetstream/iterator_test.go [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+	}
+
+	tests := []struct {
+		name     string
+		fields   fields
+		fillFunc func(chan *nats.Msg)
+		want     bool
+	}{
+		{
+			name: "true, one message",
```
[the remark this answers is not in the document]
Comment (alyakimenko) [reply source/jetstream/iterator_test.go:38]: Added

File: source/jetstream/iterator.go [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+// messageToRecord converts a *nats.Msg to a sdk.Record.
+func (i *Iterator) messageToRecord(msg *nats.Msg) (sdk.Record, error) {
+	position, err := i.getMessagePosition(msg)
+	if err != nil {
+		return sdk.Record{}, fmt.Errorf("get position: %w", err)
+	}
+
+	// retrieve a message metadata one more time to grab a metadata.Timestamp
+	// and use it for a sdk.Record.CreatedAt
+	metadata, err := msg.Metadata()
```
[the remark this answers is not in the document]
Comment (alyakimenko) [reply source/jetstream/iterator.go:233]: The `MsgMetadata` struct looks like this:
```go
// MsgMetadata is the JetStream metadata associated with received messages.
type MsgMetadata struct {
	Sequence     SequencePair
	NumDelivered uint64
	NumPending   uint64
	Timestamp    time.Time
	Stream       string
	Consumer     string
	Domain       string
}
```

We use the `Sequence` and `Timestamp` from it. And a `Stream` we get from the Iterator's `consumerInfo` field. These fields are pretty much enough to construct a unique position.

File: source/source.go [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+	"fmt"
+	"strings"
+
+	"github.com/conduitio-labs/conduit-connector-nats-jetstream/source/jetstream"
+	sdk "github.com/conduitio/conduit-connector-sdk"
+	"github.com/nats-io/nats.go"
+)
+
+// Iterator defines an iterator interface.
+type Iterator interface {
```
[the remark this answers is not in the document]
Comment (alyakimenko) [reply source/source.go]: Yeah, I agree with this. I've removed the interface

File: source/jetstream/position_test.go [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+				Subject: "foo_subject",
+				OptSeq:  32,
+			},
+			wantErr: false,
+		},
+		{
+			name: "success, empty",
+			args: args{
+				sdkPosition: sdk.Position([]byte(
+					`{"durable":"","stream":"","subject":"","opt_seq":0}`,
```
[the remark this answers is not in the document]
Comment (alyakimenko) [reply source/jetstream/position_test.go]: Changed to the `{}`

File: source/jetstream/iterator.go [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+	conn          *nats.Conn
+	messages      chan *nats.Msg
+	unackMessages []*nats.Msg
+	jetstream     nats.JetStreamContext
+	consumerInfo  *nats.ConsumerInfo
+	subscription  *nats.Subscription
+}
+
+// IteratorParams contains incoming params for the NewIterator function.
+type IteratorParams struct {
```
[the remark this answers is not in the document]
Comment (alyakimenko) [reply source/jetstream/iterator.go:45]: We would get an import cycle in this case

File: README.md [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+### Testing
+
+Run `make test` to run all the unit and integration tests, which require Docker and Docker Compose to be installed and running. The command will handle starting and stopping docker containers for you.
+
+## Source
+
+### Connection and authentication
+
+The NATS JetStream connector connects to a NATS server or a cluster with the required parameters `urls`, `subject` and `mode`. If your NATS server has a configured authentication you can pass an authentication details in the connection URL. For example, for a token authentication the url will look like: `nats://mytoken@127.0.0.1:4222`, and for a username/password authentication: `nats://username:password@127.0.0.1:4222`. But if your server is using [NKey](https://docs.nats.io/using-nats/developer/connecting/nkey) or [Credentials file](https://docs.nats.io/using-nats/developer/connecting/creds) for authentication you must configure them via seperate [configuration](#configuration) parameters.
```
[the remark this answers is not in the document]
Comment (alyakimenko) [reply README.md]: Fixed

File: README.md [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+- (optional) [golangci-lint](https://github.com/golangci/golangci-lint) 1.45.2
+- [NATS](https://nats.io/download/) 2.8.4 and [nats.go](https://github.com/nats-io/nats.go) library v1.16.0
+
+### How to build it
+
+Run `make`.
+
+### Testing
+
+Run `make test` to run all the unit and integration tests, which require Docker and Docker Compose to be installed and running. The command will handle starting and stopping docker containers for you.
```
[the remark this answers is not in the document]
Comment (alyakimenko) [reply README.md:20]: Yeah, okay, I've mentioned Docker and docker-compose in the prerequisites

File: README.md [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+
+### Connection and authentication
+
+The NATS JetStream connector connects to a NATS server or a cluster with the required parameters `urls`, `subject` and `mode`. If your NATS server has a configured authentication you can pass an authentication details in the connection URL. For example, for a token authentication the url will look like: `nats://mytoken@127.0.0.1:4222`, and for a username/password authentication: `nats://username:password@127.0.0.1:4222`. But if your server is using [NKey](https://docs.nats.io/using-nats/developer/connecting/nkey) or [Credentials file](https://docs.nats.io/using-nats/developer/connecting/creds) for authentication you must configure them via separate [configuration](#configuration) parameters.
+
+### Receiving messages
+
+The connector creates a durable NATS consumer which means it's able to read messages that were written to a NATS stream before the connector was created, unless configured otherwise. The `deliverPolicy` configuration parameter allows you to control this behavior.
+
+- If the `deliverPolicy` is equal to `new` the connector will only consume messages which were created after the connector.
```
Comment (hariso) [inline README.md:32]: > which were created after the connector.

I believe we should add `after the connector started`, to make it clear.

Comment (alyakimenko) [reply README.md]: Fixed in the https://github.com/conduitio-labs/conduit-connector-nats-jetstream/pull/5/commits/3c798bf55b81ba4077ef5bf0460b262f23162811

File: source/jetstream/iterator.go (hunk 1 above)
Comment (alyakimenko) [reply source/jetstream/iterator.go:69]: Added the async errors handler in the https://github.com/conduitio-labs/conduit-connector-nats-jetstream/pull/5/commits/d0a67bbc82040d3011e993ba3470348ede079d94

File: source/source_test.go [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+			},
+			wantErr: false,
+		},
+		{
+			name: "fail, empty config",
+			args: args{
+				ctx: context.Background(),
+				cfg: map[string]string{},
+			},
+			wantErr: true,
```
[the remark this answers is not in the document]
Comment (alyakimenko) [reply source/source_test.go]: Added in the https://github.com/conduitio-labs/conduit-connector-nats-jetstream/pull/5/commits/78b41e5e2901699bc0ec0d7cdf67ca1e500154fd

File: source/jetstream/iterator.go [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+		opts = append(opts, nats.AckNone())
+	}
+
+	opts = append(opts, []nats.SubOpt{
+		nats.Durable(params.Durable),
+		nats.ReplayInstant(),
+		nats.DeliverSubject(fmt.Sprintf("%s.%s", params.Durable, params.Stream)),
+		nats.EnableFlowControl(),
+		nats.IdleHeartbeat(heartbeatTimeout),
+	}...)
```
Comment (hariso) [inline source/jetstream/iterator.go:199]: Is there a special reason why we put the options in a slice, and then use the ellipsis operator? i.e. why not have:
```go
	opts = append(opts, 
		nats.Durable(params.Durable),
		nats.ReplayInstant(),
		nats.DeliverSubject(fmt.Sprintf("%s.%s", params.Durable, params.Stream)),
		nats.EnableFlowControl(),
		nats.IdleHeartbeat(heartbeatTimeout),
	)
```
?

File: source/source_integration_test.go [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+				}
+				t.Fatalf("read message: %v", err)
+
+				return
+			}
+
+			break
+		}
+
+		if !reflect.DeepEqual(record.Payload.Bytes(), []byte(`{"level": "info"}`)) {
```
[the remark this answers is not in the document]
Comment (alyakimenko) [reply source/source_integration_test.go:101]: Yeah, the `bytes.Equal` would be a bit faster, I just did it out of habit, thanks!

File: source/source.go [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+}
+
+// Open opens a connection to NATS and initializes iterators.
+func (s *Source) Open(ctx context.Context, position sdk.Position) error {
+	opts, err := s.getConnectionOptions()
+	if err != nil {
+		return fmt.Errorf("get connection options: %w", err)
+	}
+
+	conn, err := nats.Connect(strings.Join(s.config.URLs, ","), opts...)
```
[the remark this answers is not in the document]
Comment (alyakimenko) [reply source/source.go]: Created the method

File: source/jetstream/position_test.go [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+	"reflect"
+	"testing"
+
+	sdk "github.com/conduitio/conduit-connector-sdk"
+)
+
+func Test_position_marshalPosition(t *testing.T) {
+	t.Parallel()
+
+	type fields struct {
```
[the remark this answers is not in the document]
Comment (alyakimenko) [reply source/jetstream/position_test.go]: Yeah, thanks, made a use of the `jetstream.position`

File: source/source_integration_test.go [hunk 15]
```diff
@@ -0,0 +1,176 @@
...
+
+	if err := source.Teardown(context.Background()); err != nil {
+		t.Fatalf("teardown source: %v", err)
+	}
+}
+
+func TestSource_Read_JetStream(t *testing.T) {
+	t.Parallel()
+
+	t.Run("success, one message", func(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (alyakimenko) [reply source/source_integration_test.go]: Okay, it makes sense, thanks

File: source/jetstream/iterator.go (hunk 11 above)
Comment (alyakimenko) [reply source/jetstream/iterator.go]: No, there's no special reason for this, I've removed the redundant slice, thank you

File: README.md [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+- If the `deliverPolicy` is equal to `all` the connector will consume all messages in a stream.
+
+The connector allows you to configure a size of a pending message buffer. If your NATS server has hundreds of thousands of messages and a high frequency of their writing, it's highly recommended to set the `bufferSize` parameter high enough (`65536` or more, depending on how much RAM you have). Otherwise, you risk getting a [slow consumers](https://docs.nats.io/running-a-nats-service/nats_admin/slow_consumers) problem.
+
+### Position handling
+
+The position during this mode contains the following fields: `durable` (a durable consumer name), `stream` (a name of a stream the consumer reading from), `subject`, `timestamp` (timestamp of a message or the time the message was read by the connector) and `opt_seq` (the position of a message in a stream).
```
[the remark this answers is not in the document]
Comment (alyakimenko) [reply README.md]: Updated the section, moved the info about fields to the position's comments

**Review by hariso — APPROVED** [review]

**Pull request closed without merging**